pints/toy/_stochastic_degradation_model.py

```diff
@@ -0,0 +1,139 @@
+    *Extends:* :class:`pints.ForwardModel`, :class:`pints.toy.ToyModel`.
+    """
+
+    def __init__(self, initial_concentration=20):
+        super(StochasticDegradationModel, self).__init__()
+        self._n0 = float(initial_concentration)
+        if self._n0 < 0:
+            raise ValueError('Initial concentration cannot be negative.')
+
+        self._interp_func = None
+        self._mol_count = []
+        self._time = []
+
+    def n_parameters(self):
+        """ See :meth:`pints.ForwardModel.n_parameters()`. """
+        return 1
+
+    def simulate(self, parameters, times):
+        """ See :meth:`pints.ForwardModel.simulate()`. """
+        if len(parameters) != self.n_parameters():
+            raise ValueError('This model should have only 1 parameter.')
+        k = parameters[0]
+
+        if k <= 0:
+            raise ValueError('Rate constant must be positive.')
+
+        times = np.asarray(times)
+        if np.any(times < 0):
+            raise ValueError('Negative times are not allowed.')
+        if self._n0 == 0:
+            return np.zeros(times.shape)
+
+        time, mol_count = self.simulate_stochastically(parameters)
+
+        # Interpolate as step function, decreasing mol_count by 1 at each
+        # reaction time point
+        self._interp_func = interp1d(time, mol_count, kind='previous')
+
+        # Compute concentration values at given time points using f1
+        # at any time beyond the last reaction, concentration = 0
+        values = self._interp_func(times[np.where(times <= max(time))])
+        zero_vector = np.zeros(len(times[np.where(times > max(time))]))
+        values = np.concatenate((values, zero_vector))
+
+        return values
+
+    def simulate_stochastically(self, parameters):
+        """ Stochastic simulation according to Gillespie algorithm"""
+        if len(parameters) != self.n_parameters():
+            raise ValueError('This model should have only 1 parameter.')
+        k = parameters[0]
+        t = 0
+        if k <= 0:
+            raise ValueError('Rate constant must be positive.')
+
+        a = self._n0
+        self._mol_count = [a]
+        self._time = [t]
+
+        # Run stochastic degradation algorithm, calculating time until next
+        # reaction and decreasing concentration by 1 at that time
+        while a > 0:
+            r = np.random.uniform(0, 1)
+            t += (1 / (a * k)) * np.log(1 / r)
+            self._time.append(t)
+            a = a - 1
+            self._mol_count.append(a)
+
+        return self._time, self._mol_count
+
+    def deterministic_mean(self, parameters, times):
+        """ Calculates deterministic mean of infinitely many stochastic
+        simulations, which follows :math:: n0*exp(-kt)"""
+        if len(parameters) != self.n_parameters():
+            raise ValueError('This model should have only 1 parameter.')
+        k = parameters[0]
+
+        if k <= 0:
+            raise ValueError('Rate constant must be positive.')
```

Comment:
From Mathematica (see email), we can calculate the variance and it turns out to me E^(-2 k t) (-1 + E^(k t)) n0. Can we add a function which returns this?

Reply:
Cool! I'll add that in